src/sql/src/plan/func.rs

```diff
@@ -349,32 +356,38 @@ impl<'a> ArgImplementationMatcher<'a> {
     /// [pgparser]: https://www.postgresql.org/docs/current/typeconv-oper.html
     pub fn select_implementation<R>(
         ident: &'a str,
+        err_string_gen: fn(&str, &[Option<ScalarType>], String) -> String,
         ecx: &'a ExprContext<'a>,
         impls: &[FuncImpl<R>],
-        args: &[Expr],
+        cexprs: Vec<CoercibleScalarExpr>,
     ) -> Result<R, failure::Error> {
         // Immediately remove all `impls` we know are invalid.
-        let l = args.len();
+        let l = cexprs.len();
         let impls = impls
             .iter()
             .filter(|i| i.params.validate_arg_len(l))
             .collect();
-        let mut m = Self { ident, ecx };
-
-        let mut exprs = Vec::new();
-        for arg in args {
-            let expr = query::plan_coercible_expr(ecx, arg)?.0;
-            exprs.push(expr);
-        }
+        let m = Self { ident, ecx };
 
-        let types: Vec<_> = exprs
+        let types: Vec<_> = cexprs
             .iter()
             .map(|e| ecx.column_type(e).map(|t| t.scalar_type))
             .collect();
 
-        let f = m.find_match(&types, impls)?;
-        let exprs = m.generate_param_exprs(exprs, &f.params)?;
-        (f.op.0)(ecx, exprs)
+        // try-catch in Rust.
```

Comment:
@benesch Is there a better version of `try/catch` you know of?

Reply:
Yes, but it's unfortunately unstable: https://doc.rust-lang.org/beta/unstable-book/language-features/try-blocks.html. AFAIK what you have is the best available pattern at the moment.